ethpm/utils/bytecode_validation.py

```diff
@@ -0,0 +1,85 @@
+import re
+
+import web3
+
+from eth_utils import remove_0x_prefix
+
+from ethpm.exceptions import BytecodeMismatch
+
+
+SWARM_HASH_PREFIX = "a165627a7a72305820"
+SWARM_HASH_SUFFIX = "0029"
+EMBEDDED_SWARM_HASH_REGEX = (
+    SWARM_HASH_PREFIX +
+    "[0-9a-zA-Z]{64}" +
+    SWARM_HASH_SUFFIX +
+    "$"
+)
+
+SWARM_HASH_REPLACEMENT = (
+    SWARM_HASH_PREFIX +
+    "<" +
+    "-" * 20 +
+    "swarm-hash-placeholder" +
+    "-" * 20 +
+    ">" +
+    SWARM_HASH_SUFFIX
+)
+
+EMPTY_BYTECODE_VALUES = {None, "0x"}
+
+
+def compare_bytecode(left: str, right: str) -> bool:
+    unprefixed_left = remove_0x_prefix(left)
+    unprefixed_right = remove_0x_prefix(right)
+
+    norm_left = re.sub(EMBEDDED_SWARM_HASH_REGEX, SWARM_HASH_REPLACEMENT, unprefixed_left)
+    norm_right = re.sub(EMBEDDED_SWARM_HASH_REGEX, SWARM_HASH_REPLACEMENT, unprefixed_right)
+
+    if len(norm_left) != len(unprefixed_left) or len(norm_right) != len(unprefixed_right):
+        raise ValueError(
+            "Invariant.  Normalized bytecodes are not the correct lengths:" +
+            "\n- left  (original)  :" +
+            left +
+            "\n- left  (unprefixed):" +
+            unprefixed_left +
+            "\n- left  (normalized):" +
+            norm_left +
+            "\n- right (original)  :" +
+            right +
+            "\n- right (unprefixed):" +
+            unprefixed_right +
+            "\n- right (normalized):" +
+            norm_right
+        )
+
+    return norm_left == norm_right
+
+
+def verify_contract_bytecode(web3: web3, expected_bytecode: str, address: str) -> None:
+    """
+    Verify that expected_bytecode matches the bytecode found at address
+    """
+    # Check that the contract has bytecode
+    if expected_bytecode in EMPTY_BYTECODE_VALUES:
+        raise ValueError(
+            "Contract instances which contain an address cannot have empty "
+            "runtime bytecode"
+        )
+
+    chain_bytecode = web3.toHex(web3.eth.getCode(address))
```

Comment:
@carver Also here, the code returned from `web3.eth.getCode(address)` always begins with `0x73<address>...`, is that expected? I'm trying to just retrieve the code stored at that address

Reply:
`getCode` will give you the runtime bytecode, I believe, which does appear to start with a `'0x73'`. Is that a match?